src/librustc_target/abi/call/mod.rs

```diff
@@ -284,21 +318,34 @@ impl<'a, Ty> TyLayout<'a, Ty> {
 
                 for i in 0..self.fields.count() {
                     if !is_union && total != self.fields.offset(i) {
-                        return None;
+                        return HomogeneousAggregate::Heterogeneous;
                     }
 
                     let field = self.field(cx, i);
+
+                    // Skip over zero-sized fields for the purposes of
+                    // this check, since they don't represent any
+                    // data.
```

Comment:
You can remove this and instead change `let mut result = None;`.

This is why I preferred the `Result<Option<_>, _>` approach because it splits the two possibilities better than a 3-variant enum.

Reply:
oops, yeah, I meant to remove that line, I don't think it has any effect now. I don't see what this has to do with the return value though

Reply:
I just figured an enum would be more "self-documenting"

Reply:
(that is, this line should just go away, I don't think it matters *here* if `Result<Option>` or whatever -- maybe elsewhere)